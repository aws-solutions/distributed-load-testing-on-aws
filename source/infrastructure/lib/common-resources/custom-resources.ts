// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, CfnCondition, CfnCustomResource, CustomResource } from "aws-cdk-lib";
import { Construct } from "constructs";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Effect, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { IBucket } from "aws-cdk-lib/aws-s3";

export interface ConsoleConfigProps {
  readonly apiEndpoint: string;
  readonly cognitoIdentityPool: string;
  readonly cognitoUserPool: string;
  readonly cognitoUserPoolClient: string;
  readonly consoleBucket: IBucket;
  readonly scenariosBucket: string;
  readonly iotEndpoint: string;
  readonly iotPolicy: string;
}

export interface SendAnonymizedMetricsCRProps {
  readonly existingVpc: string;
  readonly sendAnonymizedUsage: string;
  readonly sendAnonymizedUsageCondition: CfnCondition;
  readonly solutionId: string;
  readonly solutionVersion: string;
  readonly uuid: string;
}

export interface TestingResourcesConfigCRProps {
  readonly taskCluster: string;
  readonly ecsCloudWatchLogGroup: string;
  readonly taskSecurityGroup: string;
  readonly taskDefinition: string;
  readonly subnetA: string;
  readonly subnetB: string;
}

export interface PutRegionalTemplateProps {
  readonly sourceCodeBucketName: string;
  readonly regionalTemplatePrefix: string;
  readonly mainStackRegion: string;
  readonly scenariosTable: string;
  readonly lambdaTaskRoleArn: string;
  readonly scenariosBucket: string;
}
export interface DetachIotPrincipalPolicyProps {
  readonly iotPolicyName: string;
}

/**
 * Distributed Load Testing on AWS Custom Resources Construct.
 * It creates a custom resource Lambda function, a solution UUID, and a custom resource to send anonymized usage.
 */
export class CustomResourcesConstruct extends Construct {
  private readonly customResourceLambda: NodejsFunction;

  constructor(scope: Construct, id: string, customResourceLambda: NodejsFunction) {
    super(scope, id);
    this.customResourceLambda = customResourceLambda;
  }

  private createCustomResource(
    id: string,
    customResourceFunction: NodejsFunction,
    props?: Record<string, unknown>
  ): CustomResource {
    return new CustomResource(this, id, {
      serviceToken: customResourceFunction.functionArn,
      properties: props,
    });
  }

  public getIotEndpoint() {
    const iotEndpoint = this.createCustomResource("GetIotEndpoint", this.customResourceLambda, {
      Resource: "GetIotEndpoint",
    });
    return iotEndpoint.getAtt("IOT_ENDPOINT").toString();
  }

  public detachIotPrincipalPolicy(props: DetachIotPrincipalPolicyProps) {
    this.createCustomResource("DetachIotPrincipalPolicy", this.customResourceLambda, {
      Resource: "DetachIotPolicy",
      IotPolicyName: props.iotPolicyName,
    });
  }

  public putRegionalTemplate(props: PutRegionalTemplateProps) {
    this.createCustomResource("PutRegionalTemplate", this.customResourceLambda, {
      Resource: "PutRegionalTemplate",
      SrcBucket: props.sourceCodeBucketName,
      SrcPath: props.regionalTemplatePrefix,
      DestBucket: props.scenariosBucket,
      MainRegionLambdaTaskRoleArn: props.lambdaTaskRoleArn,
      ScenariosTable: props.scenariosTable,
      MainRegionStack: props.mainStackRegion,
    });
  }

  public consoleConfig(props: ConsoleConfigProps) {
    const awsExports = `const awsConfig = {
      aws_iot_endpoint: '${props.iotEndpoint}',
      aws_iot_policy_name: '${props.iotPolicy}',
      cw_dashboard: 'https://console.aws.amazon.com/cloudwatch/home?region=${Aws.REGION}#dashboards:',
      ecs_dashboard: 'https://${Aws.REGION}.console.aws.amazon.com/ecs/home?region=${Aws.REGION}#/clusters/${Aws.STACK_NAME}/tasks',
      aws_project_region: '${Aws.REGION}',
      aws_cognito_region: '${Aws.REGION}',
      aws_cognito_identity_pool_id: '${props.cognitoIdentityPool}',
      aws_user_pools_id: '${props.cognitoUserPool}',
      aws_user_pools_web_client_id: '${props.cognitoUserPoolClient}',
      oauth: {},
      aws_cloud_logic_custom: [
          {
              name: 'dlts',
              endpoint: '${props.apiEndpoint}',
              region: '${Aws.REGION}'
          }
      ],
      aws_user_files_s3_bucket: '${props.scenariosBucket}',
      aws_user_files_s3_bucket_region: '${Aws.REGION}',
  }`;
    this.customResourceLambda.role?.addToPrincipalPolicy(
      new PolicyStatement({
        actions: ["s3:PutObject"],
        resources: [`${props.consoleBucket.bucketArn}/*`],
        effect: Effect.ALLOW,
      })
    ); // need put permission to upload config with deployment specific values
    this.createCustomResource("ConsoleConfig", this.customResourceLambda, {
      AwsExports: awsExports,
      DestBucket: props.consoleBucket.bucketName,
      Resource: "ConfigFile",
    });
  }

  public uuidGenerator() {
    const uuid = this.createCustomResource("CustomResourceUuid", this.customResourceLambda, {
      Resource: "UUID",
    });
    return {
      uuid: uuid.getAttString("UUID").toString(),
      suffix: uuid.getAttString("SUFFIX").toString(),
    };
  }

  public sendAnonymizedMetricsCR(props: SendAnonymizedMetricsCRProps) {
    const sendAnonymizedMetrics = this.createCustomResource("AnonymizedMetric", this.customResourceLambda, {
      existingVPC: props.existingVpc,
      Region: Aws.REGION,
      Resource: "AnonymizedMetric",
      SolutionId: props.solutionId,
      UUID: props.uuid,
      Version: props.solutionVersion,
    });
    const cfnSendAnonymizedMetrics = sendAnonymizedMetrics.node.defaultChild as CfnCustomResource;
    cfnSendAnonymizedMetrics.cfnOptions.condition = props.sendAnonymizedUsageCondition;
  }

  public testingResourcesConfigCR(props: TestingResourcesConfigCRProps) {
    const testingResourcesConfig = {
      region: Aws.REGION,
      subnetA: props.subnetA,
      subnetB: props.subnetB,
      ecsCloudWatchLogGroup: props.ecsCloudWatchLogGroup,
      taskSecurityGroup: props.taskSecurityGroup,
      taskDefinition: props.taskDefinition,
      taskImage: `${Aws.STACK_NAME}-load-tester`,
      taskCluster: props.taskCluster,
    };
    this.createCustomResource("TestingResourcesConfig", this.customResourceLambda, {
      TestingResourcesConfig: testingResourcesConfig,
      Resource: "TestingResourcesConfigFile",
    });
  }
}
