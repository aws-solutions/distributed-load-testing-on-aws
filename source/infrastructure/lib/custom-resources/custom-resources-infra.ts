// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { ArnFormat, CfnResource, Duration, Stack } from "aws-cdk-lib";
import { Effect, Policy, PolicyDocument, PolicyStatement, Role, ServicePrincipal } from "aws-cdk-lib/aws-iam";
import { Code, Function as LambdaFunction, Runtime } from "aws-cdk-lib/aws-lambda";
import { Bucket, IBucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";

export interface CustomResourceInfraConstructProps {
  readonly cloudWatchPolicy: Policy;
  readonly consoleBucketArn?: string;
  readonly mainStackRegion: string;
  readonly metricsUrl: string;
  readonly scenariosS3Bucket: string;
  readonly scenariosTable: string;
  readonly solutionId: string;
  readonly solutionVersion: string;
  readonly sourceCodeBucket: IBucket;
  readonly sourceCodePrefix: string;
  readonly stackType: string;
}

/**
 * Distributed Load Testing on AWS Custom Resources Construct.
 * It creates a custom resource Lambda function.
 */
export class CustomResourceInfraConstruct extends Construct {
  public customResourceArn: string;

  constructor(scope: Construct, id: string, props: CustomResourceInfraConstructProps) {
    super(scope, id);

    const sourceBucket = props.sourceCodeBucket;
    const sourceBucketArn = sourceBucket.arnForObjects("*");

    const scenariosBucket = Bucket.fromBucketName(this, "ScenariosBucket", props.scenariosS3Bucket);
    const scenariosBucketObjectArn = scenariosBucket.arnForObjects("*");

    const scenariosTableArn = Stack.of(this).formatArn({
      service: "dynamodb",
      region: props.mainStackRegion,
      resource: "table",
      resourceName: props.scenariosTable,
    });

    const customResourceRole = new Role(this, "CustomResourceLambdaRole", {
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      inlinePolicies: {
        CustomResourcePolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["s3:GetObject"],
              resources: [sourceBucketArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["s3:PutObject", "s3:DeleteObject"],
              resources: [scenariosBucketObjectArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["dynamodb:PutItem", "dynamodb:DeleteItem"],
              resources: [scenariosTableArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:DescribeEndpoint", "iot:DetachPrincipalPolicy"],
              resources: ["*"],
            }),
            new PolicyStatement({
              actions: ["iot:ListTargetsForPolicy"],
              effect: Effect.ALLOW,
              resources: [
                Stack.of(this).formatArn({
                  service: "iot",
                  resource: "policy",
                  resourceName: "*",
                  arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
                }),
              ],
            }),
          ],
        }),
      },
    });

    customResourceRole.attachInlinePolicy(props.cloudWatchPolicy);
    const cfnCustomResourceRole = customResourceRole.node.defaultChild as CfnResource;
    cfnCustomResourceRole.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W11",
          reason: "iot:DescribeEndpoint and iot:DetachPrincipalPolicy cannot specify the resource.",
        },
      ],
    });

    const customResourceLambda = new LambdaFunction(this, "CustomResourceLambda", {
      description: "CFN Lambda backed custom resource to deploy assets to s3",
      handler: "index.handler",
      role: customResourceRole,
      code: Code.fromBucket(sourceBucket, `${props.sourceCodePrefix}/${props.stackType}-custom-resource.zip`),
      runtime: Runtime.NODEJS_18_X,
      timeout: Duration.seconds(120),
      environment: {
        METRIC_URL: props.metricsUrl,
        SOLUTION_ID: props.solutionId,
        VERSION: props.solutionVersion,
        MAIN_REGION: props.mainStackRegion,
        DDB_TABLE: props.scenariosTable,
        S3_BUCKET: props.scenariosS3Bucket,
      },
    });

    if (props.stackType === "main") {
      customResourceLambda.addToRolePolicy(
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["s3:PutObject"],
          resources: [`${props.consoleBucketArn}`, `${props.consoleBucketArn}/*`],
        })
      );
    }

    this.customResourceArn = customResourceLambda.functionArn;

    const customResource = customResourceLambda.node.defaultChild as CfnResource;
    customResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W58",
          reason: "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
        },
        {
          id: "W89",
          reason: "VPC not needed for lambda",
        },
        {
          id: "W92",
          reason: "Does not run concurrent executions",
        },
      ],
    });
  }
}
