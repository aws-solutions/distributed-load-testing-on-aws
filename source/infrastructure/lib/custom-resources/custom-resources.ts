// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, CfnCondition, CfnCustomResource, CustomResource } from "aws-cdk-lib";
import { Construct } from "constructs";

export interface ConsoleConfigProps {
  readonly apiEndpoint: string;
  readonly cognitoIdentityPool: string;
  readonly cognitoUserPool: string;
  readonly cognitoUserPoolClient: string;
  readonly consoleBucketName: string;
  readonly scenariosBucket: string;
  readonly sourceCodeBucketName: string;
  readonly sourceCodePrefix: string;
  readonly iotEndpoint: string;
  readonly iotPolicy: string;
}

export interface CopyConsoleFilesProps {
  readonly consoleBucketName: string;
  readonly scenariosBucket: string;
  readonly sourceCodeBucketName: string;
  readonly sourceCodePrefix: string;
}

export interface SendAnonymizedMetricsCRProps {
  readonly existingVpc: string;
  readonly sendAnonymizedUsage: string;
  readonly sendAnonymizedUsageCondition: CfnCondition;
  readonly solutionId: string;
  readonly solutionVersion: string;
  readonly uuid: string;
}

export interface TestingResourcesConfigCRProps {
  readonly taskCluster: string;
  readonly ecsCloudWatchLogGroup: string;
  readonly taskSecurityGroup: string;
  readonly taskDefinition: string;
  readonly subnetA: string;
  readonly subnetB: string;
  readonly uuid: string;
}

export interface PutRegionalTemplateProps {
  readonly sourceCodeBucketName: string;
  readonly regionalTemplatePrefix: string;
  readonly mainStackRegion: string;
  readonly apiServicesLambdaRoleName: string;
  readonly resultsParserRoleName: string;
  readonly scenariosBucket: string;
  readonly scenariosTable: string;
  readonly taskRunnerRoleName: string;
  readonly taskCancelerRoleName: string;
  readonly taskStatusCheckerRoleName: string;
  readonly uuid: string;
}

export interface DetachIotPrincipalPolicyProps {
  readonly iotPolicyName: string;
}

export interface CustomResourcesConstructProps {
  readonly customResourceLambdaArn: string;
}

/**
 * Distributed Load Testing on AWS Custom Resources Construct.
 * It creates a custom resource Lambda function, a solution UUID, and a custom resource to send anonymized usage.
 */
export class CustomResourcesConstruct extends Construct {
  private customResourceLambdaArn: string;

  constructor(scope: Construct, id: string, props: CustomResourcesConstructProps) {
    super(scope, id);
    this.customResourceLambdaArn = props.customResourceLambdaArn;
  }

  private createCustomResource(
    id: string,
    customResourceFunctionArn: string,
    props?: Record<string, unknown>
  ): CustomResource {
    return new CustomResource(this, id, {
      serviceToken: customResourceFunctionArn,
      properties: props,
    });
  }

  public getIotEndpoint() {
    const iotEndpoint = this.createCustomResource("GetIotEndpoint", this.customResourceLambdaArn, {
      Resource: "GetIotEndpoint",
    });
    return iotEndpoint.getAtt("IOT_ENDPOINT").toString();
  }

  public detachIotPrincipalPolicy(props: DetachIotPrincipalPolicyProps) {
    this.createCustomResource("DetachIotPrincipalPolicy", this.customResourceLambdaArn, {
      Resource: "DetachIotPolicy",
      IotPolicyName: props.iotPolicyName,
    });
  }

  public putRegionalTemplate(props: PutRegionalTemplateProps) {
    this.createCustomResource("PutRegionalTemplate", this.customResourceLambdaArn, {
      Resource: "PutRegionalTemplate",
      SrcBucket: props.sourceCodeBucketName,
      SrcPath: props.regionalTemplatePrefix,
      DestBucket: props.scenariosBucket,
      APIServicesLambdaRoleName: props.apiServicesLambdaRoleName,
      MainStackRegion: props.mainStackRegion,
      ResultsParserRoleName: props.resultsParserRoleName,
      ScenariosTable: props.scenariosTable,
      TaskRunnerRoleName: props.taskRunnerRoleName,
      TaskCancelerRoleName: props.taskCancelerRoleName,
      TaskStatusCheckerRoleName: props.taskStatusCheckerRoleName,
      Uuid: props.uuid,
    });
  }

  public copyConsoleFiles(props: CopyConsoleFilesProps) {
    this.createCustomResource("CopyConsoleFiles", this.customResourceLambdaArn, {
      DestBucket: props.consoleBucketName,
      ManifestFile: "console-manifest.json",
      Resource: "CopyAssets",
      SrcBucket: props.sourceCodeBucketName,
      SrcPath: `${props.sourceCodePrefix}/console`,
    });
  }

  public consoleConfig(props: ConsoleConfigProps) {
    const awsExports = `const awsConfig = {
      aws_iot_endpoint: '${props.iotEndpoint}',
      aws_iot_policy_name: '${props.iotPolicy}',
      cw_dashboard: 'https://console.aws.amazon.com/cloudwatch/home?region=${Aws.REGION}#dashboards:',
      ecs_dashboard: 'https://${Aws.REGION}.console.aws.amazon.com/ecs/home?region=${Aws.REGION}#/clusters/${Aws.STACK_NAME}/tasks',
      aws_project_region: '${Aws.REGION}',
      aws_cognito_region: '${Aws.REGION}',
      aws_cognito_identity_pool_id: '${props.cognitoIdentityPool}',
      aws_user_pools_id: '${props.cognitoUserPool}',
      aws_user_pools_web_client_id: '${props.cognitoUserPoolClient}',
      oauth: {},
      aws_cloud_logic_custom: [
          {
              name: 'dlts',
              endpoint: '${props.apiEndpoint}',
              region: '${Aws.REGION}'
          }
      ],
      aws_user_files_s3_bucket: '${props.scenariosBucket}',
      aws_user_files_s3_bucket_region: '${Aws.REGION}',
  }`;
    this.createCustomResource("ConsoleConfig", this.customResourceLambdaArn, {
      AwsExports: awsExports,
      DestBucket: props.consoleBucketName,
      Resource: "ConfigFile",
    });
  }

  public uuidGenerator() {
    const uuid = this.createCustomResource("CustomResourceUuid", this.customResourceLambdaArn, {
      Resource: "UUID",
    });
    return {
      uuid: uuid.getAttString("UUID").toString(),
      suffix: uuid.getAttString("SUFFIX").toString(),
    };
  }

  public sendAnonymizedMetricsCR(props: SendAnonymizedMetricsCRProps) {
    const sendAnonymizedMetrics = this.createCustomResource("AnonymizedMetric", this.customResourceLambdaArn, {
      existingVPC: props.existingVpc,
      Region: Aws.REGION,
      Resource: "AnonymizedMetric",
      SolutionId: props.solutionId,
      UUID: props.uuid,
      VERSION: props.solutionVersion,
    });
    const cfnSendAnonymizedMetrics = sendAnonymizedMetrics.node.defaultChild as CfnCustomResource;
    cfnSendAnonymizedMetrics.cfnOptions.condition = props.sendAnonymizedUsageCondition;
  }

  public testingResourcesConfigCR(props: TestingResourcesConfigCRProps) {
    const testingResourcesConfig = {
      region: Aws.REGION,
      subnetA: props.subnetA,
      subnetB: props.subnetB,
      ecsCloudWatchLogGroup: props.ecsCloudWatchLogGroup,
      taskSecurityGroup: props.taskSecurityGroup,
      taskDefinition: props.taskDefinition,
      taskImage: `${Aws.STACK_NAME}-load-tester`,
      taskCluster: props.taskCluster,
    };
    this.createCustomResource("TestingResourcesConfig", this.customResourceLambdaArn, {
      TestingResourcesConfig: testingResourcesConfig,
      Resource: "TestingResourcesConfigFile",
      Uuid: props.uuid,
    });
  }
}
