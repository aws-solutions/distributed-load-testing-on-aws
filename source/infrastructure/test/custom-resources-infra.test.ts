// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Template } from "aws-cdk-lib/assertions";
import { App, DefaultStackSynthesizer, Stack } from "aws-cdk-lib";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { Policy, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { CustomResourceInfraConstruct } from "../lib/custom-resources/custom-resources-infra";

test("DLT API Test", () => {
  const app = new App();
  const stack = new Stack(app, "DLTStack", {
    synthesizer: new DefaultStackSynthesizer({
      generateBootstrapVersionRule: false,
    }),
  });
  const testSourceBucket = new Bucket(stack, "testSourceCodeBucket");
  const testPolicy = new Policy(stack, "TestPolicy", {
    statements: [
      new PolicyStatement({
        resources: ["*"],
        actions: ["cloudwatch:Get*"],
      }),
    ],
  });

  new CustomResourceInfraConstruct(stack, "TestCustomResourceInfra", {
    cloudWatchPolicy: testPolicy,
    consoleBucketArn: "test:console:bucket:arn",
    mainStackRegion: "test-region-1",
    metricsUrl: "http://testurl.com",
    scenariosS3Bucket: "scenariotestbucket",
    scenariosTable: "scenarioTestTable",
    solutionId: "S0XXX",
    solutionVersion: "testVersion",
    sourceCodeBucket: testSourceBucket,
    sourceCodePrefix: "test/source/prefix",
    stackType: "main",
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();
  Template.fromStack(stack).hasResourceProperties("AWS::Lambda::Function", {
    Code: {
      S3Bucket: {
        Ref: "testSourceCodeBucketC577B176",
      },
      S3Key: "test/source/prefix/main-custom-resource.zip",
    },
    Description: "CFN Lambda backed custom resource to deploy assets to s3",
    Environment: {
      Variables: {
        DDB_TABLE: "scenarioTestTable",
        MAIN_REGION: "test-region-1",
        METRIC_URL: "http://testurl.com",
        S3_BUCKET: "scenariotestbucket",
        SOLUTION_ID: "S0XXX",
        VERSION: "testVersion",
      },
    },
    Handler: "index.handler",
    Role: {
      "Fn::GetAtt": ["TestCustomResourceInfraCustomResourceLambdaRole03671AE8", "Arn"],
    },
    Runtime: "nodejs18.x",
    Timeout: 120,
  });
});
