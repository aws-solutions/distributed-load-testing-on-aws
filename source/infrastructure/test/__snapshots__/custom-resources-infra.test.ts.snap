// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DLT API Test 1`] = `
{
  "Resources": {
    "TestCustomResourceInfraCustomResourceLambda92E822A7": {
      "DependsOn": [
        "TestCustomResourceInfraCustomResourceLambdaRoleDefaultPolicy46B2072F",
        "TestCustomResourceInfraCustomResourceLambdaRole03671AE8",
      ],
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W58",
              "reason": "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
            },
            {
              "id": "W89",
              "reason": "VPC not needed for lambda",
            },
            {
              "id": "W92",
              "reason": "Does not run concurrent executions",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "testSourceCodeBucketC577B176",
          },
          "S3Key": "test/source/prefix/main-custom-resource.zip",
        },
        "Description": "CFN Lambda backed custom resource to deploy assets to s3",
        "Environment": {
          "Variables": {
            "DDB_TABLE": "scenarioTestTable",
            "MAIN_REGION": "test-region-1",
            "METRIC_URL": "http://testurl.com",
            "S3_BUCKET": "scenariotestbucket",
            "SOLUTION_ID": "S0XXX",
            "VERSION": "testVersion",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestCustomResourceInfraCustomResourceLambdaRole03671AE8",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestCustomResourceInfraCustomResourceLambdaRole03671AE8": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W11",
              "reason": "iot:DescribeEndpoint and iot:DetachPrincipalPolicy cannot specify the resource.",
            },
          ],
        },
      },
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "s3:GetObject",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        {
                          "Fn::GetAtt": [
                            "testSourceCodeBucketC577B176",
                            "Arn",
                          ],
                        },
                        "/*",
                      ],
                    ],
                  },
                },
                {
                  "Action": [
                    "s3:PutObject",
                    "s3:DeleteObject",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":s3:::scenariotestbucket/*",
                      ],
                    ],
                  },
                },
                {
                  "Action": [
                    "dynamodb:PutItem",
                    "dynamodb:DeleteItem",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":dynamodb:test-region-1:",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":table/scenarioTestTable",
                      ],
                    ],
                  },
                },
                {
                  "Action": [
                    "iot:DescribeEndpoint",
                    "iot:DetachPrincipalPolicy",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
                {
                  "Action": "iot:ListTargetsForPolicy",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":iot:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":policy/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "CustomResourcePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestCustomResourceInfraCustomResourceLambdaRoleDefaultPolicy46B2072F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:PutObject",
              "Effect": "Allow",
              "Resource": [
                "test:console:bucket:arn",
                "test:console:bucket:arn/*",
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestCustomResourceInfraCustomResourceLambdaRoleDefaultPolicy46B2072F",
        "Roles": [
          {
            "Ref": "TestCustomResourceInfraCustomResourceLambdaRole03671AE8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestPolicyCC05E598": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "cloudwatch:Get*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestPolicyCC05E598",
        "Roles": [
          {
            "Ref": "TestCustomResourceInfraCustomResourceLambdaRole03671AE8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testSourceCodeBucketC577B176": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
